app/static/js/ajax-port-installations.js

```diff
@@ -0,0 +1,109 @@
+function pageStateToURL(days, first, second, third) {
+    var expr = "?days=" + days + "&first=" + first + "&second=" + second + "&third=" + third;
+    history.pushState(null, null, expr)
+
+}
+
+function runPortsSortAjax(page, order_by_1, order_by_2, order_by_3, days, search_by) {
+    var currentSortAjaxRequest = null;
+    currentSortAjaxRequest = $.ajax({
+            type: 'GET',
+            url: '/statistics/ports/filter',
+            data: {
+                'page': page,
+                'order_by_1': order_by_1,
+                'order_by_2': order_by_2,
+                'order_by_3': order_by_3,
+                'days': days,
+                'search_by': search_by,
+                'csrfmiddlewaretoken': $("input[name=csrfmiddlewaretoken]").val()
+            },
+            success: portsSortAjaxSuccess,
+            beforeSend: function () {
+                $('#list-of-ports').html('');
+                $('#loading-image').show();
+                pageStateToURL(days, order_by_1, order_by_2, order_by_3);
+                if(currentSortAjaxRequest != null) {
+                    currentSortAjaxRequest.abort();
+                }
+            },
+            dataType: 'html'
+        });
+}
+
+function portsSortAjaxSuccess(data, textStatus, jqXHR) {
+    $('#list-of-ports').html($(data).filter('#response-table').html());
+    $('#navigation').html($(data).filter('#navigation-source').html());
+    $('#loading-image').hide();
+}
+
+function getPageState() {
+    var order_by_1 = $('.order_by_1').attr('id');
+    var order_by_2 = $('.order_by_2').attr('id');
+    var order_by_3 = $('.order_by_3').attr('id');
+    var days = $('#days-filter').val();
+    var search_by = $('#search-by').val();
+    return {
+        order_by_1: order_by_1,
+        order_by_2: order_by_2,
+        order_by_3: order_by_3,
+        days: days,
+        search_by: search_by
+    }
+}
+
+$(function () {
+    $('#table-header').ready(function () {
+        var state = getPageState();
+        runPortsSortAjax(1, state.order_by_1, state.order_by_2, state.order_by_3, state.days, state.search_by);
+    });
+});
+
+
+function sort(e) {
```

Comment:
I will check on this but is there no way to doing this some other way? maybe a Django package? I am sure no one would like to modify this later if there's a need.

Reply:
I searched for other ways of doing this. But none was a good way to handle large number of ports that we have. There are also packages in JQuery, but again they will first load all the objects and then perform the operations.

There is no package to achieve efficiently the suggestion mentioned here: https://github.com/macports/macports-webapp/issues/39, so I went with writing custom code.

Presence of ajax makes it even more hard to follow.